processing/src/main/java/org/apache/druid/segment/nested/ConstantColumnAndIndexSupplier.java

```diff
@@ -0,0 +1,339 @@
+import org.apache.druid.segment.column.ColumnIndexSupplier;
+import org.apache.druid.segment.column.ColumnType;
+import org.apache.druid.segment.column.DictionaryEncodedStringValueIndex;
+import org.apache.druid.segment.column.DictionaryEncodedValueIndex;
+import org.apache.druid.segment.column.DruidPredicateIndex;
+import org.apache.druid.segment.column.LexicographicalRangeIndex;
+import org.apache.druid.segment.column.NullValueIndex;
+import org.apache.druid.segment.column.NullableTypeStrategy;
+import org.apache.druid.segment.column.NumericRangeIndex;
+import org.apache.druid.segment.column.SimpleImmutableBitmapIndex;
+import org.apache.druid.segment.column.StringValueSetIndex;
+import org.apache.druid.segment.column.ValueType;
+import org.apache.druid.segment.data.BitmapSerdeFactory;
+import org.apache.druid.segment.data.VByte;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Objects;
+import java.util.SortedSet;
+
+public class ConstantColumnAndIndexSupplier implements Supplier<NestedCommonFormatColumn>, ColumnIndexSupplier
+{
+  public static ConstantColumnAndIndexSupplier read(
+      ColumnType logicalType,
+      BitmapSerdeFactory bitmapSerdeFactory,
+      ByteBuffer bb
+  )
+  {
+    final byte version = bb.get();
+    final int columnNameLength = VByte.readInt(bb);
+    final String columnName = StringUtils.fromUtf8(bb, columnNameLength);
+
+    if (version == NestedCommonFormatColumnSerializer.V0) {
+      try {
+        final int rowCount = VByte.readInt(bb);
+        final int valueLength = VByte.readInt(bb);
+        final Object constantValue = NestedDataComplexTypeSerde.OBJECT_MAPPER.readValue(
+            IndexMerger.SERIALIZER_UTILS.readBytes(bb, valueLength),
+            Object.class
+        );
+        return new ConstantColumnAndIndexSupplier(logicalType, bitmapSerdeFactory, rowCount, constantValue);
+      }
+      catch (IOException ex) {
+        throw new RE(ex, "Failed to deserialize V%s column [%s].", version, columnName);
+      }
+    } else {
+      throw new RE("Unknown version " + version);
+    }
+  }
+
+  private final ColumnType logicalType;
+  private final BitmapSerdeFactory bitmapSerdeFactory;
+  @Nullable
+  private final Object constantValue;
+
+  private final ImmutableBitmap matchBitmap;
+  private final SimpleImmutableBitmapIndex match;
+  private final ImmutableBitmap noMatchBitmap;
+  private final SimpleImmutableBitmapIndex noMatch;
+
+  public ConstantColumnAndIndexSupplier(
+      ColumnType logicalType,
+      BitmapSerdeFactory bitmapSerdeFactory,
+      int numRows,
+      @Nullable Object constantValue
+  )
+  {
+    this.logicalType = logicalType;
+    this.bitmapSerdeFactory = bitmapSerdeFactory;
+    this.constantValue = constantValue;
+    this.matchBitmap = bitmapSerdeFactory.getBitmapFactory().complement(
+        bitmapSerdeFactory.getBitmapFactory().makeEmptyImmutableBitmap(),
+        numRows
+    );
```

Comment:
Remind me of the lifecycle of the `ConstantColumnAndIndexSupplier`, I'm worried that it exists on-heap on segment-load, where I don't think it's good to pay the overhead of this on-heap object.  It would be better to just build it on-demand at query time (it's not really expensive to build).